geoh5py/groups/group_type.py

```diff
@@ -69,44 +71,45 @@ def allow_delete_content(self) -> bool:
 
     @allow_delete_content.setter
     def allow_delete_content(self, allow: bool):
+        if not isinstance(allow, bool) and allow != 1 and allow != 0:
+            raise TypeError("'allow_delete_content must be a boolean.")
         self._allow_delete_content = bool(allow)
 
+    @staticmethod
+    def create_custom(workspace: Workspace, **kwargs) -> GroupType:
+        """Creates a new instance of GroupType for an unlisted custom Group type with a
+        new auto-generated UUID.
+        """
+        return GroupType(workspace, **kwargs)
+
     @classmethod
-    def find_or_create(
-        cls, workspace: workspace.Workspace, entity_class, **kwargs
-    ) -> GroupType:
-        """Find or creates an EntityType with given UUID that matches the given
+    def find_or_create(cls, workspace: Workspace, **kwargs) -> GroupType:
+        """
+        Find or creates an EntityType with given uid that matches the given
         Group implementation class.
 
+        It is expected to have a single instance of EntityType in the Workspace
+        for each concrete Entity class.
+
+        To find an object, the kwargs must contain an existing 'uid' keyword,
+        or a 'entity_class' keyword, containing an object class.
+
         :param workspace: An active Workspace class
-        :param entity_class: An Group implementation class.
 
         :return: A new instance of GroupType.
         """
-        uid = uuid.uuid4()
-        if getattr(entity_class, "default_type_uid", None) is not None and isinstance(
-            entity_class.default_type_uid(), uuid.UUID
+        kwargs = cls.convert_kwargs(kwargs)
+        if (
+            getattr(kwargs.get("entity_class", None), "default_type_uid", None)
+            is not None
         ):
-            uid = entity_class.default_type_uid()
-
-            if "ID" in kwargs:
-                kwargs["ID"] = uid
-            else:
+            uid = kwargs["entity_class"].default_type_uid()
+            if uid is not None:
                 kwargs["uid"] = uid
-        else:
-            for key, val in kwargs.items():
-                if key.lower() in ["id", "uid"]:
-                    uid = uuid.UUID(val)
-
-        entity_type = cls.find(workspace, uid)
-        if entity_type is not None:
-            return entity_type
+        uid = kwargs.get("uid", None)
+        if uid is not None:
```

Comment:
We can probably reverse the order of operation to make it simpler.

- first check that uid is provided - if not replace with the entity_class default.
- call the `find` method to see if it exists already and return
- otherwise create and return